lib/cli/test.js

```diff
@@ -1,21 +1,191 @@
+	return remote.connect(function(loginErr, conn) {
+		if (loginErr) {
+			return onError('Login failed for remote ' + remote.name + ': ' + loginErr);
+		}
+		return onSuccess(conn);
+	});
+};
+
+var runTestsSynchronous = function (connection, classNames, verbose, callback) {
+	var conn = connection;
+
+	return conn.tooling.runTestsSynchronous(classNames, function(err, res) {
+		if (err) { return callback(err); }
+
+		var overview = formatting.formatOverview(res);
+		var messages = [];
+
+		if (verbose) {
+			messages = messages.concat(formatting.formatSuccesses(res));
+		}
+
+		messages = messages.concat(formatting.formatFailures(res));
+
+		logToConsole(messages);
+
+		if (res['failures'].length > 0) { return callback(new Error(overview)); }
+
+		return callback(null, overview);
+	});
+};
+
+var testClassNames = function(command, classNames, callback) {
+	command.opts = command.opts ? command.opts : command.docopt();
+
+  var config = new(require('../config'))();
+	// see https://github.com/jsforce/jsforce/issues/493
+  var apiVersion = command.opts['--apiVersion'] || '36.0'
+  config.set('defaultApiVersion', apiVersion);
+
+	var remote;
+	try {
+		remote = command.project.remotes.get(command.opts['<remote>']);
+	} catch (err) {
+		return callback(err);
+	}
+
+	printStartingMessage(remote);
+
+	var verbose = command.opts['--verbose'] ? true : false
+	var whenLoggedIn = function(conn) {
+		return runTestsSynchronous(conn, classNames, verbose, callback);
+	};
+	var onLoginError = function(errMessage) {
+		return callback(errMessage);
+	};
+
+	return login(remote, whenLoggedIn, onLoginError);
+};
+
+var printTestAllResults = function(res, verbose) {
+	var deployDetails = res.details;
+	var runTestResult = deployDetails.runTestResult;
+	var messages = [];
+
+	if (verbose) {
+		messages = messages.concat(formatting.formatSuccesses(runTestResult));
+	}
+
+	messages = messages.concat(formatting.formatFailures(runTestResult));
+	messages = messages.concat(formatting.formatComponentFailures(deployDetails));
+	messages = messages.concat(formatting.formatCodeCoverageWarnings(runTestResult));
+
+	return logToConsole(messages);
+}
+
+var deployEmptyCheckOnly = function(connection, archive, verbose, callback) {
+	var conn = connection;
+	var deployOpts = {
 		rollbackOnError: true,
 		checkOnly: true,
 		testLevel: 'RunLocalTests'
 	};
+	var messages = [];
+
+	conn.metadata.pollTimeout = config.pollTimeout;
+
+	return conn.metadata.deploy(archive, deployOpts).complete({
+		details: true
+	}, function(err, res) {
+		if (err) { return callback(err); }
+
+		printTestAllResults(res, verbose);
+
+		if (res.id) {
+			messages.push('Visit ' + conn.instanceUrl + '/changemgmt/monitorDeploymentsDetails.apexp?asyncId=' + res.id + ' for more information.');
+		}
+
+		if (res.status === 'Failed') {
+			if (res.details && res.details.runTestResult) {
+				messages.push(formatting.formatOverview(res.details.runTestResult));
```

Comment:
It would be good to always see the overview.
Move this before the `if (res.status === 'Failed') {`?